### ISSUE TITLE
Not Authorized Error Response on API Call

### ISSUE THREAD

**Issue by cychilds** [department-of-veterans-affairs/vets-api-clients#45]
## Email address

<cy@cychilds.com>

## APIs affected (if applicable)

<veteran_status.read>

## How can we help?

<Receiving json resonse "{"errors":[{"title":"Not authorized","detail":"Not authorized","code":"401","status":"401"}]}">

## What have you tried?

<Making the request as documented with the access_token received via the Open ID / OAuth process>

## Additional context

<VA Oauth Response (some id values below have been replaced with the field name)
Time: 1554090453

Raw Authorization Approval Information
[access_token]=(eyJraWQiOiJib2dGS2JqQ0c3VDc0UWh1ZnRiZVFwR3RxenVp...848)
[token_type]=(Bearer)
[expires_at]=(1554094052)
[scope]=(openid veteran_status.read)
[id_token]=(eyJraWQiOiJib2dGS2JqQ0c3VDc0UWh1ZnRiZVFwR3RxenVp...846)
[state]=(2)

JWT Access Token Header
[kid]=(kid)
[alg]=(RS256)

JWT Access Token Data
[ver]=(1)
[jti]=(AT.byYX-U3OiJl6AmosQX6GNwNVj6oGEavWqzKB8PDzuRw)
[iss]=(https://deptva-eval.okta.com/oauth2/default)
[aud]=(api://default)
[iat]=(1554090452)
[exp]=(1554094052)
[cid]=(cid)
[uid]=(uid)
[scp]=(array)
[sub]=(sub)

JWT ID Token Header
[kid]=(kid)
[alg]=(RS256)

JWT ID Token Data
[sub]=(00u2jpw2kx2ONsmva2p7)
[ver]=(1)
[iss]=(https://deptva-eval.okta.com/oauth2/default)
[aud]=(aud)
[iat]=(1554090452)
[exp]=(1554094052)
[jti]=(ID.rrMdP9xiVUEVjJlKotZe_NkFEXRtx52C1iKD8fS4zcw)
[amr]=(array)
[idp]=(idp)
[auth_time]=(1554090442)
[at_hash]=(_r6wxdS-PlTL7B9r_EHeTA)

Submit request to https://dev-api.va.gov/services/veteran_verification/v0/status with HTTP header 'Authorization: Bearer access_token

Receive
{"errors":[{"title":"Not authorized","detail":"Not authorized","code":"401","status":"401"}]}>

**Comment by cychilds** [comment]
I think I know what the problem may be.
All of my service center content (both inbound and outbound) is managed by a single process and that process is the "referrer" of the api call. That process name will not match my api registered redirect_url and the va api is probably rejecting my api call due to that mismatch.
What is the process to change my registered redirect_url or to add another url?

**Comment by nfasulo** [comment]
Hello @cychilds, I can update your redirect_url. What should the new value be?

**Comment by cychilds** [comment]
Thanks.
https://www.dsacentral.net/Portals/dsphp001.php

**Comment by nfasulo** [comment]
I have added https://www.dsacentral.net/Portals/dsphp001.php as a new redirect_url. Give it a try when you have a chance. Thanks.

**Comment by cychilds** [comment]
The new redirect_url is working well, but ... the end result remains the same. The flow works as documented up to the point where I reveive the access_token but any api calls therefter return the same error ... Not authorized ... status code 401. In order to move forward I just need to know why the va api considers my request unauthorized.
Is my developer account limited in some way?
Is my request malformed in some manner?
How can I get more specific information so that I can address the problem?
Your assistance would be greatly appreciated.

**Comment by cychilds** [comment]
i am now begining to think that the issue may be with the setup of my cURL configuration prior to the va api data calls I am making. Stated another way, I think the error message I am getting is being generated by cURL and is not coming from the va api.
If you could just verify that my developer account has access to the test data, I will then drill down into the cURL calls.
Thanks.

**Comment by nfasulo** [comment]
You do indeed have access to test data for veterans_verification endpoints, so I would agree that the problem is probably in you cURL configuration.

**Comment by cychilds** [comment]
I have confirmed that cURL is posting the GET request properly and the "unauthorized" message is coming from the VA API.

Below are three sections:
  1. A string of return codes from various cURL functions I call ... all return codes are 0 (OK)
  2. My request header to retrieve the veteran status
  3. The VA API response header

Using this information, could you determine why my data request is being rejected as "unauthorized"?

Thank you, sir.

Error Track: 0-0-0-0-0-0-0-0 

Request Header: GET /services/veteran_verification/v0/status HTTP/1.1
    Authorization: Bearer access_code_here
    Accept: application/json
    Host: dev-api.va.gov
    Referer: https://www.dsacentral.net/Portals/dsphp001.php

Reponse Header: HTTP/1.1 401 Unauthorized
    Date: Wed, 03 Apr 2019 02:26:24 GMT
    Content-Type: application/json; charset=utf-8
    Content-Length: 93
    Connection: keep-alive
    Access-Control-Allow-Origin: *
    Vary: Origin
    Via: kong/0.14.1
    WWW-Authenticate: Key realm="kong"
    WWW-Authenticate: Token realm="Application"
    X-Git-SHA: 749984e839bf8d5a5ee4d8217380bcdfbaa7870a
    X-GitHub-Repository: https://github.com/department-of-veterans-affairs/vets-api
    X-Kong-Proxy-Latency: 118
    X-Kong-Upstream-Latency: 489
    X-RateLimit-Limit-minute: 60
    X-RateLimit-Remaining-minute: 57
    X-Request-Id: 1bf63642-1c1c-43f3-a431-fcb8b8be1e12
    X-Runtime: 0.487432
    Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
    Set-Cookie: TS016f4012=01c8917e48bfdf1408a5322770517df6bb5cfba592de5085ba1952b5582ee0c228f6edd2da77ba11f4fd70fec5d923b1697f646cac; Max-Age=900; Path=/

Raw Data: {"errors":[{"title":"Not authorized","detail":"Not authorized","code":"401","status":"401"}]}

**Comment by cychilds** [comment]
Nick,

Just touching base ... standing by for feedback ...

**Comment by nfasulo** [comment]
Sorry, Cy, I meant to check in yesterday and forgot. I am actually getting a 401 right now as well. We are digging in to identify what the problem is. Thank you for your patience.

**Comment by nfasulo** [comment]
All right, we uncovered the problem. The mock data for the /services/veteran_verification/v0/status has been updated and I am able to make this request now (I used test account va.api.user+idme.001@gmail.com). Will you give your request another try and see if you get an good response?

**Comment by cychilds** [comment]
I am now getting the expected result. Thanks for your help and my thanks to any others that participated in the fix.
Before I close out this issue, could you tell me if other api data sources on the test platform will need to be updated and also if all test users (002 - 006) are good to go? the only reason I ask is that I don't want to report this issue if you are still working on it but I also don't want to ignore it if you think that all is well.

**Comment by nfasulo** [comment]
I did fix the other other test users as well.
The problem was due to a change to which id was being used to lookup an individual, but we didn't have mock data in place for the new lookup method. So any APIs that use the old method will still work with the original mocked data and any that use the new lookup method now have the data they need to move forward.
Thanks again for your patience as we worked through this problem.
